Add Optimism Bedrock

kaladinlight opened the pull request:
- Optimism support for bedrock upgrade
- Closing https://github.com/trezor/blockbook/pull/868 in favor of this PR and considering upgrade _should_ be coming up within the next month
  - removes local node dependencies for building the data-transport-layer
  - removes alchemy L2 rpc url dependency

Blockbook:
![image](https://user-images.githubusercontent.com/35275952/227645266-9f7bebb9-d2c1-48e2-a8c7-03f568b7949c.png)

Etherscan:
![image](https://user-images.githubusercontent.com/35275952/227645317-468813ef-0e33-4ecd-a99c-64c73d12d933.png)

kaladinlight commented on api/types.go:
```diff
@@ -232,6 +232,10 @@ type EthereumSpecific struct {
 	GasLimit          *big.Int                               `json:"gasLimit"`
 	GasUsed           *big.Int                               `json:"gasUsed"`
 	GasPrice          *Amount                                `json:"gasPrice"`
+	L1Fee             *big.Int                               `json:"l1Fee,omitempty"`
+	L1FeeScalar       string                                 `json:"l1FeeScalar,omitempty"`
+	L1GasPrice        *Amount                                `json:"l1GasPrice,omitempty"`
+	L1GasUsed         *big.Int                               `json:"l1GasUsed,omitempty"`
```
L1 fee data returned on optimism transaction receipts (needed to calculate total transaction fee L2 + L1)

kaladinlight commented on bchain/coins/eth/ethrpc.go:
```diff
@@ -795,12 +798,13 @@ func (b *EthereumRPC) GetTransactionForMempool(txid string) (*bchain.Tx, error)
-	var tx *bchain.RpcTransaction
+	tx := &bchain.RpcTransaction{}
 	hash := ethcommon.HexToHash(txid)
-	err := b.RPC.CallContext(ctx, &tx, "eth_getTransactionByHash", hash)
+	err := b.RPC.CallContext(ctx, tx, "eth_getTransactionByHash", hash)
 	if err != nil {
 		return nil, err
-	} else if tx == nil {
+	}
+	if *tx == (bchain.RpcTransaction{}) {
```
Initialize and handle empty struct case to avoid `**bchain.RpcTransaction` type

kaladinlight commented on bchain/coins/eth/ethtx.proto:
```diff
@@ -1,30 +1,34 @@
+    bytes GasUsed = 1;
+    bytes Status = 2;
+    repeated LogType Log = 3;
+    optional bytes L1Fee = 4;
```
As discussed in past PR's, proto type is backwards compatible

kaladinlight commented on bchain/coins/optimism/evm.go:
```diff
@@ -0,0 +1,45 @@
+		return err
+	}
+
+	// special case to handle empty gas price for a valid rpc transaction
+	// (https://goerli-optimism.etherscan.io/tx/0x9b62094073147508471e3371920b68070979beea32100acdc49c721350b69cb9)
+	if r, ok := result.(*bchain.RpcTransaction); ok {
+		if *r != (bchain.RpcTransaction{}) && r.GasPrice == "" {
+			r.GasPrice = "0x0"
+		}
+	}
```
Unknown if this will be a scenario on mainnet, but it is valid on testnet. I added here just to try and keep the edge case scoped as locally as possible

kaladinlight commented on build/templates/backend/scripts/optimism.sh:
```diff
@@ -0,0 +1,43 @@
+DATA_DIR={{.Env.BackendDataPath}}/{{.Coin.Alias}}/backend
+
+CHAINDATA_DIR=$DATA_DIR/geth/chaindata
+SNAPSHOT=
```
Update when mainnet upgrade occurs and snapshots are released.

kaladinlight commented on build/templates/backend/scripts/optimism.sh:
```diff
@@ -0,0 +1,43 @@
+  --ws.api=eth,net,web3,debug,txpool,engine \
+  --ws.origins="*" \
+  --rollup.disabletxpoolgossip=true \
+  --rollup.sequencerhttp=https://sequencer.optimism.io \
```
Verify this sequencer address is accurate when mainnet upgrade occurs

kaladinlight commented on build/templates/backend/scripts/optimism_archive.sh:
```diff
@@ -0,0 +1,47 @@
+DATA_DIR={{.Env.BackendDataPath}}/{{.Coin.Alias}}/backend
+
+CHAINDATA_DIR=$DATA_DIR/geth/chaindata
+SNAPSHOT=
```
ditto

kaladinlight commented on build/templates/backend/scripts/optimism_archive.sh:
```diff
@@ -0,0 +1,47 @@
+  --ws.origins="*" \
+  --rollup.disabletxpoolgossip=true \
+  --rollup.historicalrpc=http://127.0.0.1:8303 \
+  --rollup.sequencerhttp=https://sequencer.optimism.io \
```
ditto

kaladinlight commented on build/templates/backend/scripts/optimism_legacy_geth.sh:
```diff
@@ -0,0 +1,40 @@
+DATA_DIR={{.Env.BackendDataPath}}/{{.Coin.Alias}}/backend
+
+CHAINDATA_DIR=$DATA_DIR/geth/chaindata
+SNAPSHOT=
```
ditto

kaladinlight commented on build/templates/backend/scripts/optimism_testnet_archive.sh:
```diff
@@ -0,0 +1,47 @@
+DATA_DIR={{.Env.BackendDataPath}}/{{.Coin.Alias}}/backend
+
+CHAINDATA_DIR=$DATA_DIR/geth/chaindata
+SNAPSHOT=https://storage.googleapis.com/oplabs-goerli-data/goerli-bedrock.tar
```
there is not an archive snapshot available and I can't find a testnet genesis file, so this isn't the most accurate. we can just remove the archive testnet if you would like, let me know your thoughts

kaladinlight commented on configs/coins/optimism.json:
```diff
@@ -0,0 +1,66 @@
+    "package_revision": "satoshilabs-1",
+    "system_user": "optimism",
+    "version": "0.11.2-13ee9ab",
+    "binary_url": "https://github.com/ethereum-optimism/op-geth/archive/refs/tags/v1.11.2-13ee9ab.tar.gz",
```
check for official release when mainnet upgrade occurs

kaladinlight commented on configs/coins/optimism_op_node.json:
```diff
@@ -0,0 +1,35 @@
+	  "verification_source": "bc71f75ef330f494e17f78c61c39296a6a66a325e634158c784d72352d25cd91",
+	  "extract_command": "mkdir backend/source && tar -C backend/source --strip 1 -xf",
+	  "exclude_files": [],
+	  "exec_command_template": "/bin/sh -c '{{.Env.BackendInstallPath}}/{{.Coin.Alias}}/op-node --l1 http://127.0.0.1:8136 --l1.trustrpc --l1.rpckind=basic --l2 http://127.0.0.1:8400 --rpc.addr 127.0.0.1 --rpc.port {{.Ports.BackendRPC}} 2>> {{.Env.BackendDataPath}}/{{.Coin.Alias}}/backend/{{.Coin.Alias}}.log'",
```
verify if `--network` flag will be needed for mainnet

kaladinlight commented on docs/ports.md:
```diff
@@ -1,78 +1,86 @@
+| eCash                        | 9097                    | 9197                  | 8097             | 38397                             |
+| Avalanche                    | 9098                    | 9198                  | 8098             | 38398 p2p                         |
+| Avalanche Archive            | 9099                    | 9199                  | 8099             | 38399 p2p                         |
+| Optimism                     | 9200                    | 9300                  | 8200             | 38400 p2p, 8300 http, 8400 auth   |
+| Optimism Op-Node             |                         |                       | 8201             |                                   |
+| Optimism Archive             | 9202                    | 9302                  | 8202             | 38402 p2p, 8302 http, 8402 auth   |
+| Optimism Legacy Geth         |                         |                       | 8203             | 38403 p2p, 8303 http              |
```
relevant diff

kaladinlight commented on docs/ports.md:
```diff
@@ -1,78 +1,86 @@
+| Flo Testnet                  | 19066                   | 19166                 | 18066            | 48366                             |
+| Qtum Testnet                 | 19088                   | 19188                 | 18088            | 48388                             |
+| Omotenashicoin Testnet       | 19089                   | 19189                 | 18089            | 48389                             |
+| Optimism Testnet             | 19200                   | 19300                 | 18200            | 48400 p2p, 18300 http, 18400 auth |
+| Optimism Testent Op-Node     |                         |                       | 18201            |                                   |
+| Optimism Testnet Archive     | 19202                   | 19302                 | 18202            | 48402 p2p, 18302 http, 18402 auth |
+| Optimism Legacy Geth Testnet |                         |                       | 18203            | 48403 p2p, 18303 http             |
```
relevant diff

kaladinlight commented:
@martinboehm ready for review, successful bedrock upgrade on my end and blockbook still working as expected

kaladinlight commented:
@martinboehm if you have any time at some point, would we be able to upstream optimism (bedrock)? I have been running the blockbook for the past 6 months or so without any issues. I am not using the infra, so that could use some :eyes: to make sure everything is working as expected. Let me know if you have time or any thoughts. Thank you!

martinboehm commented:
@kaladinlight Hi. I finally have time to check this PR. Sorry for the delay.
I have a problem to pull the PR from your repository as an anonymous user (I can do it with my account). Is this on purpose?

kaladinlight commented:
> @kaladinlight Hi. I finally have time to check this PR. Sorry for the delay. I have a problem to pull the PR from your repository as an anonymous user (I can do it with my account). Is this on purpose?

Hey! Good to hear from you. Wasn't sure I would haha. Hmm... Not on purpose, just normal fork repo to my knowledge. I just looked at settings and nothing jumped out at me.

kaladinlight commented:
Also since you are back to look, I will take the time to update the infra with current versions, etc.

martinboehm commented:
@kaladinlight thank you! 
In the meantime, I actually realized that it is my fault, sorry. I used ssh link instead of https. I can access the repo without a problem.
And I was able to build Blockbook. Haven't tried to build backends yet.

kaladinlight commented:
> @kaladinlight thank you! In the meantime, I actually realized that it is my fault, sorry. I used ssh link instead of https. I can access the repo without a problem. And I was able to build Blockbook. Haven't tried to build backends yet.

Ok cool, give me a second to update it so we only need to test once and I will give you a ping when ready.

kaladinlight commented:
@martinboehm ok, pushed some updates for the backend. note I haven't ever actually run the archive setup fully, but feel free to test it and happy to help debug any issues

martinboehm commented:
@kaladinlight Blockbook runs well. Just be aware of this commit https://github.com/trezor/blockbook/commit/35ab7a3966964376100e77b4768c1c7d9a258b3c. Maybe it would be a good idea to backport it to this PR.

I successfully built `backend-optimism-archive-op-node.service`. However when I started it, it exited with this report:
```
/bin/sh: 1: Syntax error: redirection unexpected
backend-optimism-archive-op-node.service: Main process exited, code=exited, status=2/INVALIDARGUMENT
```
After I modified the redirection to `2>&1 >>` and tried again, I got the following error:
```
/bin/sh: 1: /opt/coins/nodes/optimism_archive_op_node/op-node: not found
```
I realized that you are doing the build as a post install step. However, there is no go environment on the server. Could you please move the build step to the `extract_command` in a similar way it is done for example in `configs/coins/polygon_heimdall.json`.

kaladinlight commented:
@martinboehm Hey, just got back from travel. Happy to hear blockbook is running well. I will go ahead and backport that commit and take a closer look at the backend config to sort those issues and give you a ping.

kaladinlight commented:
@martinboehm I updated the op-node config to build during the extract stage and updated the redirect syntax. I ran things locally (manual exec command vs service setup due to my env) and things were where they should be and started as expected. I did not have all local services (optimism/legacy geth) to fully initialize and start syncing, just verified it was running and trying to connect to things. Let me know if that fixes the issues on your side!

martinboehm commented:
@kaladinlight Thank you, the reported issues are fixed. However, run into another. The `op-node` is not starting, with this error:
```
failed to setup: unable to create the rollup node config: failed to load p2p config: failed to load p2p priv key: failed to store new p2p priv key: open opnode_p2p_priv.txt: permission denied
```
It is probably tryinig to store the `opnode_p2p_priv.txt` somewhere where it does not have permissions. It should store all files to the `/opt/coins/data/optimism_archive_op_node/backend` folder or subfolders, if possible.

kaladinlight commented:
@martinboehm There doesn't look to be any directory flag for the op_node local files. I just updated the exec command to run from the backend dir to ensure those files are created and stored there as expected.

martinboehm commented:
@kaladinlight Thank you, unfortunately I get `backend-optimism-archive-op-node.service: Failed to locate executable cd: No such file or directory` 🤦‍♂️

kaladinlight commented:
@martinboehm bleh... will take a look at a different strategy.

kaladinlight commented:
@martinboehm ok, I took another look and did find some path flags for these files, they were just nested under `p2p` unlike the typical singular `dir` or `datadir` setup. will be updating to use those and then things should be sorted finally. thanks for you patience

kaladinlight commented:
@martinboehm Ok, pushed those updates. Hopefully the last thing. Let me know if everything is working as expected now, thanks!

martinboehm commented:
@kaladinlight you are amazing, thank you 👍 `op-node` is starting now. Now I tried to build `backend-optimism-archive` and there is the same problem - the go build is post step instead of in build step. Could you please change it too?

kaladinlight commented:
@martinboehm moved the go build into the extract command. I kept the `openssl`  command in the postinst script as I would think that should included by default on any host system (can change if you would like, lemme know).

martinboehm commented:
Hmm, build was without problem, however on start I get
```
Resolving datadirs.optimism.io (datadirs.optimism.io)... 162.19.233.77
Connecting to datadirs.optimism.io (datadirs.optimism.io)|162.19.233.77|:443... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: https://datadirs.optimism.io/latest/ [following]
--2024-07-01 21:51:35--  https://datadirs.optimism.io/latest/
Reusing existing connection to datadirs.optimism.io:443.
HTTP request sent, awaiting response... 200 OK
Length: 129 [text/html]
Saving to: <E2><80><98>STDOUT<E2><80><99>

     0K                                                       100%  424M=0s

2024-07-01 21:51:35 (424 MB/s) - written to stdout [129/129]

zstd: /*stdin*\: unsupported format
tar: This does not look like a tar archive
tar: Exiting with failure status due to previous errors
```
It is because `wget https://datadirs.optimism.io/latest` is not returning the zst file but instead a small file containing
```
<html><head><meta http-equiv="refresh" content="0; url=https://datadirs.optimism.io/mainnet-2024-06-08.tar.zst" /></head></html>
```

kaladinlight commented:
@martinboehm can you give it another try with this update f0729e2378654ad06daabe4f847a114acf95a055

martinboehm commented:
It looks good now 👍

martinboehm commented:
@kaladinlight  I have probably the last request/question. I would prefer `OP` to be the shortcut for Optimism instead of `"shortcut": "ETH"`

kaladinlight commented:
> @kaladinlight I have probably the last request/question. I would prefer `OP` to be the shortcut for Optimism instead of `"shortcut": "ETH"`

The shortcut is used for the native asset symbol which is accurately `ETH` on optimism. `OP` is their native token, but not the native gas asset.

![image](https://github.com/trezor/blockbook/assets/35275952/881f3874-8497-4acb-bb5e-6c358af6cd79)

kaladinlight commented:
@martinboehm just making sure you have seen the above comment since I didn't @ you directly. Let me know if that makes sense, or if you have any questions there. Looking forward to getting this in!

blackbird-sarodh commented:
qq: will this chain type be compatible with Base Layer 2?  Sorry if not relevant. Thanks!

kaladinlight commented:
> qq: will this chain type be compatible with Base Layer 2? Sorry if not relevant. Thanks!

This specific PR only supports optimism, but I do have base support working and coming next based off this. Stay tuned!

martinboehm commented:
@kaladinlight Thank you!!
